Added PyThreadPool as wrapper around MlirLlvmThreadPool in MLIR python bindings

vfdev-5 opened the pull request:
Description:
- Exposed MlirLlvmThreadPool as PyThreadPool in MLIR Python bindings
- Add tests

Context:

In JAX ir.Context are used with disabled multi-threading to avoid caching multiple threading pools:
https://github.com/jax-ml/jax/blob/623865fe9538100d877ba9d36f788d0f95a11ed2/jax/_src/interpreters/mlir.py#L606-L611
However, when context has enabled multithreading it also uses locks on the StorageUniquers and this can be helpful to avoid data races in the multi-threaded execution (for example with free-threaded cpython, https://github.com/jax-ml/jax/issues/26272).
With this PR user can enable the multi-threading: 1) enables additional locking and 2) set a shared threading pool such that cached contexts can have one global pool.


cc @hawkinsp

vfdev-5 committed:
Added PyThreadPool as wrapper around MlirLlvmThreadPool in MLIR python bindings

vfdev-5 committed:
Added a test

joker-eph commented on mlir/test/python/ir/context_lifecycle.py:
```diff
@@ -47,3 +47,17 @@
+tp = mlir.ir.ThreadPool()
+assert tp.get_max_concurrency() > 0
+c5 = mlir.ir.Context()
+c5.enable_multithreading(False)
```
Is this needed? What happens if you don't set it?

joker-eph replied:
That's what I feared, then we need to defend against it in the Python API. That is we shouldn't crash and instead either throw an exception when the user didn't disable multi-threading first, or do it for the user implicitly in the call to `set_thread_pool()`

joker-eph replied:
We can also take it as optional at construction indeed!

But can't there be cases where you handed over an existing context and this "set_thread_pool" API can be also useful?

hawkinsp replied:
I'm just wondering if disabling multithreading and changing thread pools are safe things to do. Maybe it's fine.

vfdev-5 replied:
We can also call internally equivalent of `enable_multithreading(False)` inside `set_thread_pool()` in the binding?

vfdev-5 committed:
Added get_num_threads and _mlir_thread_pool_ptr methods to _BaseContext
Added thread_pool arg to the constructor: `mlir.ir.Context(thread_pool=tp)`

vfdev-5 commented:
@joker-eph in the last PR I added thread_pool arg to `mlir.ir.Context` constructor. Also exposed `get_num_threads` and `_mlir_thread_pool_ptr` methods:
- updated MLIR C API: added mlirContextGetNumThreads, mlirContextGetThreadPool methods

`_mlir_thread_pool_ptr` is an internal method to ensure that C++ thread pool is correctly set up.

Let me know if this works. Thanks!

joker-eph commented on mlir/python/mlir/_mlir_libs/__init__.py:
```diff
@@ -148,13 +148,18 @@ def process_initializer_module(module_name):
+                if thread_pool is None:
+                    self.enable_multithreading(True)
+                else:
+                    self.set_thread_pool(thread_pool)
```
We ignore the thread_pool argument if `disable_multithreading` is set: can we throw instead?

vfdev-5 replied:
Sounds good, I'll commit this:
```
    class Context(ir._BaseContext):
        def __init__(
            self, load_on_create_dialects=None, thread_pool=None, *args, **kwargs
        ):
            super().__init__(*args, **kwargs)
            self.append_dialect_registry(get_dialect_registry())
            for hook in post_init_hooks:
                hook(self)
            if disable_multithreading and thread_pool is not None:
                raise ValueError(
                    "Context constructor has given thread_pool argument, "
                    "but disable_multithreading flag is True. "
                    "Please, set thread_pool argument to None or "
                    "set disable_multithreading flag to False."
                )
            if not disable_multithreading:
```
If there are any other suggestions, please let me know

joker-eph reviewed (approved):
LG with one minor thing to check.

vfdev-5 committed:
Raise error if disable_multithreading and thread_pool is given
